Set output evaluation context for a single node

somebody1234 opened the pull request:
### Pull Request Description
- Closes #8072
  - Implement handlers for the corresponding buttons on the circular menu
  - Add missing icons and styles
  - Add functionality to match and extract ASTs

### Important Notes
None

### Screenshots
![image](https://github.com/enso-org/enso/assets/4046547/9d937df2-d138-4b06-b760-fa75719eeb78)
![image](https://github.com/enso-org/enso/assets/4046547/d87fbb92-c9d4-4588-b654-77caf86f9c9c)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

somebody1234 committed:
`isMatch` function to match CST subtrees

somebody1234 committed:
(broken) Detect current Output Context enabled state

somebody1234 committed:
Implement `extractMatches` to extract CST subtrees

somebody1234 committed:
Interactivity for output context override in Circular Menu; initial implementation for AST replacement

Untested

The matched segments also need to be hidden to, but we need to make sure that the propagated content edits have the correct offset.

somebody1234 committed:
Basic tests for string escape

somebody1234 committed:
Attempt to hide overridden output context expressions from `GraphNode`

somebody1234 committed:
Merge branch 'develop' into wip/sb/set-eval-ctx

somebody1234 committed:
Fix CSS for `CircularMenu`

somebody1234 committed:
Fix AST match extraction by using `parseLine` instead of `parse`

somebody1234 committed:
Fix inserted text

somebody1234 committed:
Refactor to make it easier to fix node icons for overriden output context in the future

vitvakatu commented:
A little issue in logic:

https://github.com/enso-org/enso/assets/6566674/901ecf4b-9318-46b8-864b-4f4ecbf13e57

When using the button near the node, context switch expressions should never stack; instead, we should replace the existing expression with the correct one. Having a “no-op” expression, like enabling output context in the “live” environment is totally okay, and should be considered as no context switch expression present on a node.

somebody1234 commented:
@vitvakatu Hmm... that would mean you can't have different overrides for different environments, right?

somebody1234 committed:
Only allow a single output context override

farmaazon commented on app/gui2/src/components/GraphEditor/GraphNode.vue:
```diff
@@ -143,6 +162,70 @@ const dragPointer = usePointer((pos, event, type) => {
   }
 })
 
+const exprWithoutOverriddenOutputContext = computed(() => {
```
The name is not precise, as it returns more than just an expression.

farmaazon commented on app/gui2/src/components/GraphEditor/GraphNode.vue:
```diff
@@ -11,15 +11,35 @@ import { injectGraphSelection } from '@/providers/graphSelection'
 
+const enableMethodName = 'with_enabled_context'
+const disableMethodName = 'with_disabled_context'
+function withOverriddenOutputContextPatternText(name: string, enable: boolean) {
```
Do we need the `PatternText` suffix here? I think the function sounds ok without it.

vitvakatu commented:
> @vitvakatu Hmm... that would mean you can't have different overrides for different environments, right?

Yes, this is how the old GUI is implemented. We can confirm with @jdunkerley to make sure. I tried to find the actual discussion around this behavior but failed (there were a lot of drastic changes to the design).

I’m sure that any operation on button shouldn’t cause the change in the visible node’s expression.

farmaazon commented on app/gui2/src/components/GraphEditor/GraphNode.vue:
```diff
@@ -143,6 +162,70 @@ const dragPointer = usePointer((pos, event, type) => {
+const isOutputContextOverridden = computed({
+  get() {
+    return (
+      exprWithoutOverriddenOutputContext.value != null &&
+      exprWithoutOverriddenOutputContext.value.enabled !== projectStore.isOutputContextEnabled &&
+      exprWithoutOverriddenOutputContext.value.contextExpr.repr().slice(1, -1) ===
+        projectStore.executionMode
```
I would make an alias for `exprWithoutOverriddenOutputContext.value` to shorten this expression somewhat. Also, the last could be a bit decomposed; the value returned from `slice` should be a semantically named variable for easy reading.

somebody1234 commented:
> I’m sure that any operation on button shouldn’t cause the change in the visible node’s expression.

@vitvakatu note that there are alternatives that still preserve this behavior - I think ideally we would recursively hide *all* context overrides, but only modify the override for the current context (or add one, if none are for the current context)

somebody1234 commented:
@Frizi thoughts on using widgets for this instead? i think it'd be a cleaner (+ more generic?) solution, *but* the issue is that we *have* to communicate the "is node overridden" data all the way up the tree back to the `GraphNode` (or via an `inject`) - which is doable, but doesn't exist yet

farmaazon commented on app/gui2/src/util/ast/string.ts:
```diff
@@ -0,0 +1,17 @@
+  '`': '``',
+}
+
+/** Escape a string so it can be safely spliced into an interpolated (`''`) string.
+ * NOT USABLE to insert into raw strings. Does not include quotes. */
```
I would mention here what exactly you mean by string - is it Enso string or JS string?

farmaazon reviewed (approved):
Any chance for output context overwriting tests? They could be E2E tests if UT would be hard, and they can be done in a separate PR.

somebody1234 committed:
Address review

somebody1234 commented:
i'll see what i can do for UT

somebody1234 committed:
Initial component test for `GraphNode` including execution mode override

Warnings about `onScopeDispose` being called outside an effect scope

Errors caused by unhandled promise rejections

somebody1234 committed:
Get rid of `unhandledRejection` errors

somebody1234 commented:
Re-requesting review from @farmaazon for feedback on UT: [Implementation](https://github.com/enso-org/enso/pull/8440/files?w=1#diff-6a631a03514b4e1bfad12f7b5a41d744477d41b27280ff324a1cc294b5ab96c4)

kazcw committed:
Inline function return signatures

kazcw committed:
More consistent handling of `default`

kazcw committed:
GUI1

farmaazon reviewed (approved):
The test, if works, is good. I'm only afraid that it's a bit complex and may be hard to maintain (the test makes a lot of mocking). But let's see - if it would be too much a burden, we may fix it later.

Personally, when I have a similar thing to test, I try to extract the logic I want to test (like the Component Browser's `Input` class). But anyway, I think we should not prolong this PR.

somebody1234 commented:
@farmaazon Yeah, that's true. But in this case, I think it's not complex enough to warrant creating a new class.

re: mocking - I agree that this is a bit much, but the idea is that we can use the mocks added(/used) here in other tests too. Of course, at that point we should refactor *all* the mocks used out somewhere else - perhaps a new subdirectory

somebody1234 committed:
Merge branch 'develop' into wip/sb/set-eval-ctx

somebody1234 commented:
@farmaazon actually, i don't think it'll hurt to prolong it a bit because it might need to be refactored to use the new AST anyway

farmaazon commented:
> re: mocking - I agree that this is a bit much, but the idea is that we can use the mocks added(/used) here in other tests too. Of course, at that point we should refactor _all_ the mocks used out somewhere else - perhaps a new subdirectory

I know this approach, we did it recently (You may find in old GUI `Mock` structures for general use), but I didn't like them. Usually they made adding new tests/cases difficult; I needed to grasp a big structure of mocks to realize what is there, and how to configure it for my case; usually I didn't find a way and tried to generalize the mocks, but that makes even more convoluted code, and at the end I had failing both new and old tests :)

> But in this case, I think it's not complex enough to warrant creating a new class.

I disagree, for me, it is a part of logic nice to extract, because it has no (or minimum) dependencies. And it will become more complex with SKIP and FREEZE macros in the future.

somebody1234 commented:
in this case i think the mocks are fine since they should all be defaults - however, it's true that refactoring things out to a new class that doesn't depend on Vue will make it much easier to understand and much less fragile.

i don't think it *hurts* to have a test depend on Vue, since if the test breaks, it's likely that there is a real regression - however i don't think it makes sense to have two tests doing the same thing, so might as well extract it out to its own function, especially considering SKIP and FREEZE macros.

somebody1234 committed:
Port AST matching to new API

kazcw committed:
Add missing issue number

kazcw committed:
Tests

somebody1234 committed:
Fix tests and stuff

`GraphNode` test is broken because `replaceExpressionContent` seems to
be broken.

somebody1234 commented:
hmm... so i guess it'll take longer than expected, but i guess it isn't too much of an issue considering the estimate is 5

(specifically: i'm currently trying to generalize it, and... well, i think the hardest part will be frankensteining multiple ASTs together, based on when a child is equal to the next one in the list. and then somehow serializing the new AST too... all while preserving whitespace)

somebody1234 committed:
WIP: Checking and modifying prefixes of node content

somebody1234 committed:
Fix tests

somebody1234 committed:
Fix bug

somebody1234 committed:
Fix imports in E2E tests

somebody1234 commented:
i think the new approach is a *huge* bandaid. but for now it works, and hopefully most of the jank can be removed once we get real AST edits?

somebody1234 commented:
oh yeah, one *minor* (okay, major) issue: for now at least, i've made the new AST types public, because i need to read and mutate them outside of `abstract.ts`

somebody1234 commented:
requesting review from @kazcw because of ast-related changes... maybe this task should be delayed until we have edits?

somebody1234 committed:
Move `ast/string` to `ast/text`

somebody1234 committed:
Fix text escaping

somebody1234 committed:
Fix error that only occurs on built package

kazcw reviewed (approved):
I will need to make some changes to the AST-editing support here to fit my planned direction for the APIs, but I think it's OK to merge; I can update it in-tree in my next PR.

kazcw committed:
Merge remote-tracking branch 'origin/wip/kw/parse-inline-signatures' into ast-edits

kazcw committed:
Merge remote-tracking branch 'origin/wip/sb/set-eval-ctx' into ast-edits

Frizi commented on app/gui2/mock/index.ts:
```diff
@@ -0,0 +1,92 @@
+  return mockNode()
+}
+
+export function waitForMainModule(projectStore?: ProjectStore) {
```
I can't find any usages of this function. Can it be removed?

somebody1234 replied:
was required previously when I used component tests for GraphNode

Frizi commented on app/gui2/mock/providers.ts:
```diff
@@ -0,0 +1,55 @@
+import { Rect } from '../src/util/rect'
+import { Vec2 } from '../src/util/vec2'
+
+export const graphNavigator: GraphNavigator = {
```
Why does this have to be mocked? I feel like no providers should need any mocking, except for things related to network requests.

somebody1234 replied:
previously we had component tests for GraphNode... I guess the correct solution is maybe to use a wrapper component then?

but also, we're no longer doing component tests so it's not like these are currently used either.

Frizi commented on app/gui2/mock/vue.ts:
```diff
@@ -0,0 +1,19 @@
+
+// It is currently not feasible to use generics here, as the type of the component's emits
+// is not exposed.
+export function handleEmit(wrapper: VueWrapper<any>, event: string, fn: (...args: any[]) => void) {
```
Is this used anywhere?

somebody1234 replied:
as above, previously used for component tests. i figured it's fine to keep it if we may want to do component testing in the future, but i have no problems with removing all unused code because we are probably not planning to do that anytime soon anyway.

Frizi commented on app/gui2/src/util/ast/match.ts:
```diff
@@ -0,0 +1,208 @@
+    placeholder,
+  )
+  if (replacements.length !== 0) {
+    // throw new Error(`${replacements.length} unused replacements.`)
```
Is this error commented out unintentionally?

somebody1234 replied:
no - nice catch, thanks, will fix.

somebody1234 commented:
For posterity: the additional logs added on errors were added because Vitest was throwing "unhandled exception" errors .

Note also that the fix that worked there, was to add mocks for the LS and data server.

kazcw committed:
wip

kazcw committed:
Merge remote-tracking branch 'origin/develop' into ast-edits

kazcw committed:
wip

kazcw committed:
Pass tests

kazcw committed:
Merge commit '0974d5ff1e1729148e7b74c98c82e03a5ad8a751' into ast-edits

kazcw committed:
Merge commit '3d76459ba7079c6bd115bbfec2b0911881e527e5' into ast-edits

kazcw committed:
Merge commit 'be2ae3279cbfddf30007f2007b0aaa19cfa14891' into ast-edits

kazcw committed:
Merge remote-tracking branch 'origin/develop' into ast-edits

somebody1234 commented on app/gui2/src/util/ast/match.ts:
```diff
@@ -0,0 +1,114 @@
+  return out
+}
+
+if (import.meta.vitest) {
```
@kazcw this should probably be in `util/ast/__tests__/match.test.ts` btw

kazcw committed:
Clean up tests

kazcw committed:
Rerun CI?

kazcw committed:
Rerun CI?

somebody1234 committed:
Merge branch 'develop' into wip/sb/set-eval-ctx
